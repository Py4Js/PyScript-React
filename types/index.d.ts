declare module "pyscript-react" {
  export type PyScriptPropertiesBase = Omit<
    ReactElementProps<
      DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>
    >,
    "children"
  >;

  export type PyScriptPropsBase = PyScriptPropertiesBase;

  export type PyScriptPropertiesWithChildren = {
    src?: never;
    children: string;
    source?: never;
  };

  export type PyScriptPropsWithChildren = PyScriptPropertiesWithChildren;

  export type PyScriptPropertiesWithSourceFullName = {
    src?: never;
    children?: never;
    source: string;
  };

  export type PyScriptPropsWithSourceFullName =
    PyScriptPropertiesWithSourceFullName;

  export type PyScriptPropertiesWithSource = {
    src: string;
    children?: never;
    source?: never;
  };

  export type PyScriptPropsWithSource = PyScriptPropertiesWithSource;

  export type PyScriptPropertiesWitOutputPart = {
    output: string;
    generateOutputTag?: boolean | keyof ReactHTML;
  };

  export type PyScriptPropsWitOutputPart = PyScriptPropertiesWitOutputPart;

  export type PyScriptPropertiesWitoutOutputPart = {
    output?: never;
    generateOutputTag?: never;
  };

  export type PyScriptPropsWitoutOutputPart =
    PyScriptPropertiesWitoutOutputPart;

  export type PyScriptPropertiesWithPyEnvPart = {
    pyEnvContent: PyEnvPropertiesBase["children"];
    pyEnvProps?: Omit<PyEnvPropertiesBase, "children">;
  };

  export type PyScriptPropsWithPyEnvPart = PyScriptPropertiesWithPyEnvPart;

  export type PyScriptPropertiesWithWithoutPyEnvPart = {
    pyEnvContent?: never;
    pyEnvProps?: never;
  };

  export type PyScriptPropsWithWithoutPyEnvPart =
    PyScriptPropertiesWithWithoutPyEnvPart;

  export type PyScriptPropertiesWithoutGeneric = PyScriptPropertiesBase &
    (PyScriptPropertiesWithPyEnvPart | PyScriptPropertiesWithWithoutPyEnvPart) &
    (PyScriptPropertiesWitOutputPart | PyScriptPropertiesWitoutOutputPart) &
    (
      | PyScriptPropertiesWithSource
      | PyScriptPropertiesWithChildren
      | PyScriptPropertiesWithSourceFullName
    );

  export type PyScriptProperties<T> = T extends infer T
    ? T & PyScriptPropertiesWithoutGeneric
    : PyScriptPropertiesWithoutGeneric;

  export type PyScriptTag = {
    <T extends object>(properties: PyScriptProperties<T>): JSX.Element;
    propTypes: WeakValidationMap<PyScriptPropertiesWithoutGeneric>;
  };

  export const PyScript: PyScriptTag;
  export const PyTitle: PyTitleTag;
  export const PyScriptProvider;
  export const PyRepl;
  export const PyRegisterWidget;
  export const PyInputBox;
  export const PyEnv;
  export const PyConfig;
  export const PyButton;
  export const PyBox;

  export default PyScript;
}
