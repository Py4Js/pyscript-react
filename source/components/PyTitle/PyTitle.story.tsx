import { ComponentMeta, ComponentStory } from "@storybook/react";
import PyScriptProvider from "../PyScriptProvider/PyScriptProvider";
import PyTitle from "./PyTitle";
import { PyTitleProperties, PyTitleTag } from "./PyTitle.types";

export default {
  title: "PyTitle",
  component: PyTitle,
} as ComponentMeta<typeof PyTitle>;

const Template: ComponentStory<PyTitleTag> = <T extends object>({
  ...rest
}: PyTitleProperties<T>): JSX.Element => {
  return (
    <PyScriptProvider>
      <py-title {...rest}>Example title</py-title>
    </PyScriptProvider>
  );
};

export const PyTitleExample: ComponentStory<typeof PyTitle> = Template.bind({});

PyTitleExample.args = {
  children: "Example title",
};
