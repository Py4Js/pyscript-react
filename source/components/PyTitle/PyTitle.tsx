import propTypes from "prop-types";
import {
  PyTitleProperties,
  PyTitleTag,
} from "~components/PyTitle/PyTitle.types";

const PyTitle: PyTitleTag = <T extends object>({
  children,
  ...rest
}: PyTitleProperties<T>): JSX.Element => {
  return <py-title {...rest}>{children}</py-title>;
};

PyTitle.propTypes = {
  children: propTypes.string.isRequired,
};

export default PyTitle;
