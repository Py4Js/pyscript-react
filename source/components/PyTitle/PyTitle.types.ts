import { DetailedHTMLProps, HTMLAttributes, WeakValidationMap } from "react";
import ReactElementProps from "~types/ReactElementProps/ReactElementProps";

export type PyTitlePropertiesBase = Omit<
  ReactElementProps<
    DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>
  >,
  "children"
> & {
  children: string;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type PyTitleProperties<T> = T extends infer T
  ? T & PyTitlePropertiesBase
  : PyTitlePropertiesBase;

export type PyTitleProps<T> = PyTitleProperties<T>;

export type PyTitleTag = {
  <T extends object>(properties: PyTitleProperties<T>): JSX.Element;
  propTypes: WeakValidationMap<PyTitlePropertiesBase>;
};
