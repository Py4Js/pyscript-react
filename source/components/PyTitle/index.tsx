import PyTitle from "~components/PyTitle/PyTitle";
import type {
  PyTitleProperties,
  PyTitlePropertiesBase,
  PyTitleProps,
  PyTitleTag,
} from "~components/PyTitle/PyTitle.types";

export { PyTitlePropertiesBase, PyTitleProperties, PyTitleProps, PyTitleTag };

export default PyTitle;
