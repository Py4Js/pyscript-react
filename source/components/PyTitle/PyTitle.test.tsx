import { render, RenderResult } from "@testing-library/react";
import { expectType } from "tsd";
import PyTitle from "~components/PyTitle/PyTitle";
import { PyTitleTag } from "~components/PyTitle/PyTitle.types";

describe("PyScript", (): void => {
  it("Renders component properly", async (): Promise<void> => {
    const { container }: RenderResult = render(
      <PyTitle>Example title</PyTitle>,
    );
    expect(container).toMatchSnapshot();
  });
  it("Has correct type", (): void => {
    expectType<PyTitleTag>(PyTitle);
  });
});
