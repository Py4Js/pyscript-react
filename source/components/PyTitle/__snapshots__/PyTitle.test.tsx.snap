// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PyScript Renders component properly 1`] = `
<div>
  <py-title>
    Example title
  </py-title>
</div>
`;
